### PR TITLE
Use HTCondor Python bindings in autoscaler

### PR DESCRIPTION
**Pull request by tpdownes**
- Replace all shell call-outs and parsing with efficient queries to the scheduler and collector daemons
- Anticipate future work to base auto-scaling on GPU and memory requests by including those job requests in queries
- Alter scaling code to account for the number of CPUs requested by each job. Previous solution implicitly assumes each job requests 1 CPU and will underscale for multi-core jobs
- Do not scale down any execute points that have not been active for less than 150 seconds

### Submission Checklist

* [x] Have you installed and run this change against pre-commit? (`pre-commit
  install`)
* [x] Are all tests passing? (`make tests`)
* [ ] Have you written unit tests to cover this change?
* [x] Is unit test coverage still above 80%?
* [x] Have you updated all applicable documentation?
* [x] Have you followed the guidelines in our Contributing document?

### CODE REVIEW

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (nick-stroud) [inline community/modules/scripts/htcondor-install/files/autoscaler.py:193]: Make "RequestCpus" a constant (all caps var), maybe other request strings as well although less important as they are not duplicated.

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (nick-stroud) [inline community/modules/scripts/htcondor-install/files/autoscaler.py:201]: Why no longer report jobs on hold?

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (nick-stroud) [inline community/modules/scripts/htcondor-install/files/autoscaler.py:197]: Will condor people understand the RequestCpus reference. For me this is less clear than the previous message (even though it is more specific) because it leaves me wondering what the reference to RequestCpus means.

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (nick-stroud) [inline community/modules/scripts/htcondor-install/files/autoscaler.py:262]: I believe `time.time()` could replace `datetime.datetime.now().timestamp()` and would be more idiomatic python.

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (nick-stroud) [inline community/modules/scripts/htcondor-install/files/autoscaler.py:189]: Silly question: What is `ads`?

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (tpdownes) [reply community/modules/scripts/htcondor-install/files/autoscaler.py:193]: A few lines above I link to HTCondor Job ClassAd attributes (for `JobStatus` and its defined values).

Would you like me to make that link more generic?

https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.html#RequestCpus

Strictly speaking, the names of ClassAd attributes in HTCondor are defined to be case insensitive (though I can't find a link to that). What I can say is that every job gets a RequestCpus (and maybe RequestMemory, RequestGpus, RequestDisk?) with a default value.

So we can do what we want, but convention is CamelCase.

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (tpdownes) [reply community/modules/scripts/htcondor-install/files/autoscaler.py:201]: The prior calculation consumed the stdout of a particular command that added up jobs in the queue. It did so in a manner that required subtracting off held jobs before calculating the autoscaling target. The new calculation computes the right target from the right data in the first place.

Jobs can go on hold deliberately or by HTCondor itself but the autoscaler isn't really the place to log that unless it's useful for autoscaling.

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (tpdownes) [reply community/modules/scripts/htcondor-install/files/autoscaler.py:197]: HTCondor admins will know `RequestCpus` (for reasons noted above) but it's easy for me to alter the text to something more completely correct: "Total CPUs requested by running/idle jobs".

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (tpdownes) [reply community/modules/scripts/htcondor-install/files/autoscaler.py:262]: I'm anticipating the need for `datetime` in future work, but I agree that `time.time()` is simpler.

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (tpdownes) [reply community/modules/scripts/htcondor-install/files/autoscaler.py:197]: I have added more descriptive comments and links that might help maintainers of this code who are less familiar with HTCondor administration.

**Comment by tpdownes** [comment]
Several changes made to address feedback. PTAL @nick-stroud

File: community/modules/scripts/htcondor-install/files/autoscaler.py
Comment (nick-stroud) [reply community/modules/scripts/htcondor-install/files/autoscaler.py:197]: My request here was to the following between L187 and L188 (could also go at top of file)

```
REQUEST_CPU = 'RequestCpus'
```

And then use that variable in place of the string literals. This prevents typos and update errors.

**Comment by tpdownes** [comment]
PTAL. Changes have been tested on a running pool.

**Review by nick-stroud — APPROVED** [review]

**Pull request merged**
